R/insert-ui.R

```diff
@@ -0,0 +1,194 @@
+#' \code{multiple} determines whether Shiny should insert the UI object
+#' relative to all matched elements or just relative to the first
+#' matched element (default).
+#'
+#' @param where Where your UI object should go relative to the selector:
+#' \describe{
+#'   \item{\code{beforeBegin}}{Before the selector element itself}
+#'   \item{\code{afterBegin}}{Just inside the selector element, before its
+#'   first child.}
+#'   \item{\code{beforeEnd}}{Just inside the selector element, after its
+#'   last child}
+#'   \item{\code{afterEnd}}{After the selector element itself (default)}
+#' }
+#' Adapted from
+#' \href{https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentHTML}{here}.
+#'
+#' @param ui The UI object you want to insert. This can be anything that
+#' you usually put inside your apps's  \code{ui} function.
+#'
+#' @param immediate Whether the UI object should be immediately inserted into
+#' the app when you call \code{insertUI}, or whether Shiny should wait for the
+#' next time your app is updated for some other reason (default).
+#'
+#' @param container A function to generate an HTML element to contain the UI
+#' object.
+#'
+#' @param inline Use an inline (\code{span()}) or block container (\code{div()},
+#' default) for the output.
+#'
+#' @param session The shiny session within which to call \code{insertUI}.
+#'
+#' @seealso \code{\link{removeUI}}
+#'
+#' @examples
+#' ## Only run this example in interactive R sessions
+#' if (interactive()) {
+#' # Define UI
+#' ui <- fluidPage(
+#'   actionButton("add", "Add UI")
+#' )
+#'
+#' # Server logic
+#' server <- function(input, output, session) {
+#'   observeEvent(input$add, {
+#'     insertUI(
+#'       selector = "#add",
+#'       ui = textInput(paste0("txt", input$add),
+#'                      "Insert some text")
+#'     )
+#'   })
+#' }
+#'
+#' # Complete app with UI and server components
+#' shinyApp(ui, server)
+#' }
+#'
+#' @export
+insertUI <- function(selector,
+  multiple = FALSE,
+  where = c("beforeBegin", "afterBegin", "beforeEnd", "afterEnd"),
+  ui,
+  immediate = FALSE,
+  container = if (inline) "span" else "div",
+  inline = FALSE,
+  session = getDefaultReactiveDomain()) {
+
+  force(selector)
+  force(ui)
+  force(session)
+  if (missing(where)) where <- "afterEnd"
+  where <- match.arg(where)
+
+  sendInsertUI <- function() {
+    session$sendCustomMessage("shiny-insert-ui",
+      list(
+        selector = selector,
+        multiple = multiple,
+        where = where,
+        content = processDeps(ui, session),
+        container = container
+      )
+    )
+  }
+
+  if (!immediate) {
+    session$onFlushed(sendInsertUI, once = TRUE)
+  } else {
+    sendInsertUI()
+  }
+}
+
+
+#' Remove UI objects
+#'
+#' Remove a UI object from the app.
+#'
+#' This function allows you to remove any part of your UI. Once you call
+#' \code{removeUI} on some element, it is gone forever.
```

Comment:
"Once you call" is fairly accurate if you pass immediate=TRUE, but the default behavior is to remove the UI at the end of the current... uh... flush cycle? I guess we don't have a name for it... anyway, the UI will be removed the next time the outputs/errors are flushed to the client.